test/e2e/petset.go

```diff
@@ -177,6 +209,175 @@ var _ = framework.KubeDescribe("PetSet [Slow] [Feature:PetSet]", func() {
+			for i, event := range Added(podTracker) {
+				if event.pod.Name != expectedOrder[i] {
+					framework.Failf("Order of scale up is wrong: %v != %v", expectedOrder[i], event.pod.Name)
+				}
+			}
+
+			By("scale down will halt with unhealthy pet")
+			pst.breakProbe(ps, testProbe)
+			pst.waitForRunningAndNotReady(3, ps)
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 0 })
+			pst.confirmPetCount(3, ps, 10*time.Second)
+
+			By("scalling down petset " + psName + " to 0 replicas and waiting until none of pods will run " + ns)
+			pst.restoreProbe(ps, testProbe)
+			pst.scale(ps, 0)
+
+			By("verifying that petset " + psName + " was scaled down in reverse order")
+			expectedOrder = []string{"pet-2", "pet-1", "pet-0"}
+			for i, event := range Deleted(podTracker) {
+				if event.pod.Name != expectedOrder[i] {
+					framework.Failf("Order of scale down is wrong: %v != %v", expectedOrder[i], event.pod.Name)
+				}
+			}
+
+			By("verifying that scaling up before scale down is finished will not be out of order")
+			pst.scale(ps, 3)
+			podTracker = []podEvent{}
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 5 })
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 2 })
+			pst.scale(ps, 4)
+			var previous string
+			for _, event := range Added(podTracker) {
+				if previous != "" {
+					msg := fmt.Sprintf("scaling should be in order, %v should be after %v", previous, event.pod.Name)
+					if strings.Split(previous, "-")[1] > strings.Split(event.pod.Name, "-")[1] {
+						framework.Failf(msg)
+					} else {
+						framework.Logf(msg)
+					}
+				}
+				previous = event.pod.Name
+			}
+		})
+
+		It("scaling down before scale up is finished will wait until current pod will be running and ready", func() {
+			By("creating petset " + psName + " in namespace " + ns)
+			testProbe := &api.Probe{Handler: api.Handler{HTTPGet: &api.HTTPGetAction{Path: "/index.html", Port: intstr.IntOrString{IntVal: 80}}}}
+			ps := newPetSet(psName, ns, headlessSvcName, 1, nil, nil, labels, testProbe)
+			ps, err := c.Apps().PetSets(ns).Create(ps)
+			pst := &petSetTester{c: c}
+			Expect(err).NotTo(HaveOccurred())
+			pst.waitForRunningAndReady(1, ps)
+
+			By("verifying that scaling down before scale up is finished will wait until current pod will be running and ready")
+			// pet-2 should be the last pet scale up will create before will start scale down
+			lastPet := "pet-2"
+			podTracker = []podEvent{}
+			pst.setHealthy(ps)
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 3 })
+			pst.waitForRunningAndReady(2, ps)
+			pst.breakProbe(ps, testProbe)
+			pst.waitForRunningAndNotReady(2, ps)
+			pst.setHealthy(ps)
+			pst.update(ps.Namespace, ps.Name, func(ps *apps.PetSet) { ps.Spec.Replicas = 1 })
+			pst.restoreProbe(ps, testProbe)
+			for i := len(podTracker) - 1; i >= 0; i-- {
+				event := podTracker[i]
```

Comment:
Could you explain more on what this loop is doing? Is it looking at the last (and only the last) event on `pet-2` to see if the pod is ready? Do we need to wait before the check?